core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,422 @@
+                    if (newValue == 0) {
+                        return null; // delete
+                    }
+                    return newValue;
+                });
+    }
+
+    private static class PendingAcquire
+    {
+        private final NodeRequirements nodeRequirements;
+        private final SettableFuture<InternalNode> future;
+
+        private PendingAcquire(NodeRequirements nodeRequirements)
+        {
+            this.nodeRequirements = requireNonNull(nodeRequirements, "nodeRequirements is null");
+            this.future = SettableFuture.create();
+        }
+
+        public NodeRequirements getNodeRequirements()
+        {
+            return nodeRequirements;
+        }
+
+        public SettableFuture<InternalNode> getFuture()
+        {
+            return future;
+        }
+
+        public long getMemoryLease()
+        {
+            return nodeRequirements.getMemory().toBytes();
+        }
+    }
+
+    private class BinPackingNodeLease
+            implements NodeAllocator.NodeLease
+    {
+        private final ListenableFuture<InternalNode> node;
+        private final AtomicBoolean released = new AtomicBoolean();
+        private final long memoryLease;
+
+        private BinPackingNodeLease(ListenableFuture<InternalNode> node, long memoryLease)
+        {
+            this.node = requireNonNull(node, "node is null");
+            this.memoryLease = memoryLease;
+        }
+
+        @Override
+        public ListenableFuture<InternalNode> getNode()
+        {
+            return node;
+        }
+
+        @Override
+        public void release()
+        {
+            if (released.compareAndSet(false, true)) {
+                node.cancel(true);
+                if (node.isDone() && !node.isCancelled()) {
+                    updateAllocatedMemory(getFutureValue(node), -memoryLease);
+                    wakeupProcessPendingAcquires();
+                }
+            }
+            else {
+                throw new IllegalStateException("Node " + node + " already released");
+            }
+        }
+    }
+
+    private static class BinPackingSimulation
+    {
+        private final InternalNodeManager.NodesSnapshot nodesSnapshot;
+        private final List<InternalNode> allNodesSorted;
+        private final Map<String, Long> nodesRemainingMemory;
+        private final ImmutableMap<String, Long> nodePoolSizes;
+
+        public BinPackingSimulation(
+                InternalNodeManager.NodesSnapshot nodesSnapshot,
+                Map<String, Long> nodePoolSizes,
+                Map<String, Long> preReservedMemory, boolean scheduleOnCoordinator)
+        {
+            this.nodesSnapshot = requireNonNull(nodesSnapshot, "nodesSnapshot is null");
+            // use same node ordering for each simulation
+            this.allNodesSorted = nodesSnapshot.getAllNodes().stream()
+                    .filter(node -> scheduleOnCoordinator || !node.isCoordinator())
```

Comment:
This should take into account coordinator only tasks (similar to FullNodeCapableAllocator)